### DISCUSSION TITLE
Order of Lifecycle Hooks and 404 pages

### DISCUSSION THREAD

**Discussion by delta1186** [livewire/livewire#3875] [1. Help] [answered]
I am a bit confused with the wording on the docs and was hoping someone could clarify the lifecycle hooks. 

https://laravel-livewire.com/docs/2.x/reference#alpine-component-object
https://laravel-livewire.com/docs/2.x/lifecycle-hooks

In my component I am getting the following order with just writing some log messages.  And just to be clear this is from a postback after the component is mounted.

hydrateFoo
boot
hydrate
my method
render
dehydrate
dehydrateFoo

So is the correct?  Working on a way to detect if the model Foo was deleted by another user and if it is I don't want my component to throw a 404 error, but instead gracefully reset and close the component.  Working with Echo to emit a FooDeleted event which I listen for on the component.  This all works, but getting the 404 when the component picks up the event.  The 404 is thrown before any of my log entries are written which I assume is because it attempts to hydrate Foo, but it no longer exists.  Kinda wish the hydrate hooks matched the tense of the update looks. ie: hydratedFoo and hydrated.  It would also be assume to have a hydratingFoo and hydrating hooks that fire before the actual hydrating.  Same for dehydrate. V3 maybe? 

I could then prevent the 404 because I could check if the model exists and pivot my code path from there.  Just thinking out loud. @calebporzio  @joshhanley  :)

When the component is initially loaded it is:
boot
mount
render
dehydrate
dehydrateFoo

```blade
    @if($ticket->id)
    <script>
        Echo.join("tickets.{{ $ticket->id }}")
            .listen('TicketUpdated', (e) => {
                window.Livewire.emit('ticketUpdated');
            })
            .listen('TicketDeleted', (e) => {
                //window.Livewire.emit('ticketDeleted');
   
                //might  have to go this route if I cannot figure out the model hydration issue when deleted,
                window.location.replace(e.url)
            });
    </script>
    @endif
```
And if you are wondering why I have to register the Echo code in my view it is because the $ticket is not initially set when the component is mounted and based on what I have read you cannot dynamically update the listeners on the component.  

So for now I am going to just take the approach shown here were I am just redirecting the user back to the same page.

**Comment by delta1186** [comment]
Just as a quick follow-up, I ended up refactoring to just not use the model as a public property and then added the following to the boot method.   Using this approach I can switch to using window.Livewire.emit('ticketDeleted') and avoid the full page reload.
```php
 public $ticketId;

 private Ticket $ticket; 

 public function boot()
    {
        if ($this->ticketId) {
            $ticket = Ticket::find($this->ticketId);
            $this->ticket = $ticket ?? Ticket::make();
        }
    }
```

**Comment by joshhanley** [comment]
@delta1186 yeah the order on hydrate isn't right, we're looking into it at the moment.

**Comment by joshhanley** [comment, marked as answer]
@delta1186 this has been fixed in PR #4026 and also a `booted` method has been added that runs after `mount` and `hydrate` in PR #4033, and both have been released in 2.7.0.

Hope this helps!
